### PR TITLE
Annotation API

### PR DESCRIPTION
**Pull request by vladak**
This change add simple annotation API.

Limitations:
  - no support for paging
  - description is HTML encoded

While testing I found out that Git annotations are broken so fix for that is included.

### CODE REVIEW

File: opengrok-indexer/src/main/java/org/opengrok/indexer/history/Annotation.java [hunk 1]
```diff
@@ -24,6 +24,7 @@
 
 package org.opengrok.indexer.history;
 
+import com.fasterxml.jackson.annotation.JsonProperty;
```
Comment (tulinkry) [inline opengrok-indexer/src/main/java/org/opengrok/indexer/history/Annotation.java]: I don't see this being used in this changeset, was it a missing import?

File: opengrok-web/src/main/java/org/opengrok/web/util/FileUtil.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+import javax.servlet.http.HttpServletResponse;
+import java.io.File;
+import java.io.IOException;
+
+public class FileUtil {
+
+    private static final RuntimeEnvironment env = RuntimeEnvironment.getInstance();
+
+    public static File getFile(String path, HttpServletResponse response) throws IOException {
```
Comment (tulinkry) [inline opengrok-web/src/main/java/org/opengrok/web/util/FileUtil.java]: Could you refactor this function to throw related exceptions and map these exceptions in exception mapper? It would make it obvious and without side effects. If I imported this function from a library I would not know what it does ... it takes a path and a response for some reason and returning a file.

Also a javadoc on public util method would not hurt.

File: apiary.apib [hunk 3]
```diff
@@ -6,6 +6,31 @@ OpenGrok RESTful API documentation. The following endpoints are accessible under
...
+
++ Parameters
+  + path (string) - path of file, relative to source root
+
++ Response 200 (application/json)
+  + Body
+
+            [
+              {
+                "number": 1,
```
Comment (tulinkry) [inline apiary.apib]: what is the number?

File: apiary.apib [hunk 4]
```diff
@@ -6,6 +6,31 @@ OpenGrok RESTful API documentation. The following endpoints are accessible under
...
+                "revision": "e2736ff6",
+                "author": "Vladimir Kotal",
+                "desc": "changeset:&nbsp;e2736ff6&lt;br/&gt;summary:&nbsp;add&nbsp;microbadges&lt;br/&gt;user:&nbsp;Vladimir&nbsp;Kotal&nbsp;&#60;vlada@devnull.cz&#62;&lt;br/&gt;date:&nbsp;Fri&nbsp;Jun&nbsp;28&nbsp;21:33:41&nbsp;CEST&nbsp;2019"
```
Comment (tulinkry) [inline apiary.apib]: can have long naming "description"

File: apiary.apib [hunk 5]
```diff
@@ -6,6 +6,31 @@ OpenGrok RESTful API documentation. The following endpoints are accessible under
...
+              },
```
Comment (tulinkry) [inline apiary.apib]: we also utilize the version number in all version of the file (like version 4 out of 22), can you include that as well?

File: apiary.apib (hunk 4 above)
Comment (tulinkry) [reply apiary.apib]: I wouldn't expect a html code inside json in a description field, is it easy to get rid of it?

File: opengrok-indexer/src/main/java/org/opengrok/indexer/history/Annotation.java (hunk 1 above)
Comment (vladak) [reply opengrok-indexer/src/main/java/org/opengrok/indexer/history/Annotation.java]: yes

File: apiary.apib (hunk 3 above)
Comment (vladak) [reply apiary.apib]: just entry number, I guess that's not necessary.

File: apiary.apib (hunk 5 above)
Comment (vladak) [reply apiary.apib]: I don't understand. Could you rephrase or provide example ?

File: opengrok-indexer/src/main/java/org/opengrok/indexer/history/GitRepository.java [hunk 6]
```diff
@@ -389,7 +389,6 @@ private String getFirstRevision(String fullpath) throws IOException {
                 ensureCommand(CMD_PROPERTY_KEY, CMD_FALLBACK),
                 "rev-list",
                 "--reverse",
-                "--max-count=1",
```
Comment (idodeclare) [inline opengrok-indexer/src/main/java/org/opengrok/indexer/history/GitRepository.java]: No big problem here but some food for thought:

* `Executor.exec()` and `Executor.exec(boolean)` actually spool the entire output of a program into memory. Probably that's not a burdensome amount with this removal of `--max-count`, given this is a `rev-list` of a single file.
* When the spooled bytes are read later at line 402, they are read with the default Java charset, which is locale dependent on the OS installation. For hashes of a `rev-list`, that is perhaps always OK — but I'm just remarking since elsewhere in `GitRepository`, specific effort is made to use a correct charset (see `newLogReader()`).
* Technically when read later at line 402, a `BufferedReader` is overkill, given that the bytes are already fully spooled in-memory.
* Indeed all the output of a `Process` (via `exec()`) must be read or else a deadlock could happen as one awaits the process exit value. The default spool does consume everything.

I wonder if this is an opportunity to create a `HeadHandler` implementation of `Executor.StreamHandler` that is line-oriented (e.g. characters out of bytes), that demands a specified encoding instead of allowing the default Java charset, and that reads and stores a maximum number of lines and then just consumes and discards bytes (without bothering with character conversion) to exhaust the stream and avoid allocating heap.

File: opengrok-web/src/main/java/org/opengrok/web/util/FileUtil.java [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+
+package org.opengrok.web.util;
+
+import org.opengrok.indexer.configuration.RuntimeEnvironment;
+
+import javax.servlet.http.HttpServletResponse;
+import java.io.File;
+import java.io.IOException;
+
+public class FileUtil {
```
Comment (idodeclare) [inline opengrok-web/src/main/java/org/opengrok/web/util/FileUtil.java]: Should the constructor be made private for a static utility class?

Comment (vladak) [reply opengrok-web/src/main/java/org/opengrok/web/util/FileUtil.java]: yep, actually the `checkstyle` complains about it.

File: apiary.apib (hunk 5 above)
Comment (tulinkry) [reply apiary.apib]: Snímek obrazovky 2020-02-27 v 17 19 50

File: opengrok-web/src/main/java/org/opengrok/web/util/FileUtil.java [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+     * @throws NoPathException
+     */
+    public static File getFile(String path, HttpServletResponse response) throws NoPathException, NoFileException {
+        if (path == null) {
+            throw new NoPathException("Missing path parameter");
+        }
+
+        File file = new File(env.getSourceRootFile(), path);
+        if (!file.isFile()) {
+            throw new NoFileException("File not found");
```
Comment (tulinkry) [inline opengrok-web/src/main/java/org/opengrok/web/util/FileUtil.java]: There is already FileNotFound (or similar) in java core

File: opengrok-web/src/main/java/org/opengrok/web/util/FileUtil.java [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+        }
+    }
+
+    /**
+     * @param path path relative to source root
+     * @return file object corresponding to the file under source root
+     * @throws NoFileException
+     * @throws NoPathException
+     */
+    public static File getFile(String path, HttpServletResponse response) throws NoPathException, NoFileException {
```
Comment (tulinkry) [inline opengrok-web/src/main/java/org/opengrok/web/util/FileUtil.java]: I think I wrote it somewhere else too, I'm little bit against this naming. But if you feel comfortable, you can disregard.

Comment (tulinkry) [inline opengrok-web/src/main/java/org/opengrok/web/util/FileUtil.java]: response is not needed

File: opengrok-indexer/src/main/java/org/opengrok/indexer/history/GitRepository.java (hunk 6 above)
Comment (tulinkry) [reply opengrok-indexer/src/main/java/org/opengrok/indexer/history/GitRepository.java]: Definitely good idea, I'd go for LineHandler (or similar) and then derive the HeadHandler out of it. For simplicity a lot of code could benefit from this.

Comment (idodeclare) [reply opengrok-indexer/src/main/java/org/opengrok/indexer/history/GitRepository.java]: Thank you, @tulinkry .

Maybe the reverse though: a `LineHandler` might be an unbounded `HeadHandler`. The latter would need to interact low-level with the stream in order to be able to read and dump bytes when a line limit is reached.

File: apiary.apib (hunk 3 above)
Comment (tulinkry) [reply apiary.apib]: you can include it but give it better name :-D

File: opengrok-web/src/main/java/org/opengrok/web/util/FileUtil.java (hunk 9 above)
Comment (vladak) [reply opengrok-web/src/main/java/org/opengrok/web/util/FileUtil.java]: toFile() ?

File: opengrok-web/src/main/java/org/opengrok/web/util/FileUtil.java (hunk 8 above)
Comment (vladak) [reply opengrok-web/src/main/java/org/opengrok/web/util/FileUtil.java]: aha ! thanks.

File: apiary.apib (hunk 5 above)
Comment (vladak) [reply apiary.apib]: Oh that. I never noticed it before :-)

File: opengrok-web/src/main/java/org/opengrok/web/util/FileUtil.java (hunk 2 above)
Comment (vladak) [reply opengrok-web/src/main/java/org/opengrok/web/util/FileUtil.java]: Makes sense, thanks for pointing out the existence of exception mappers.

File: apiary.apib (hunk 3 above)
Comment (vladak) [reply apiary.apib]: it would make sense to have it if pagination was supported.

File: apiary.apib (hunk 4 above)
Comment (vladak) [reply apiary.apib]: it turns out it is.

File: opengrok-web/src/main/java/org/opengrok/web/api/v1/NoPathExceptionMapper.java [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+
+import javax.ws.rs.core.Response;
+import javax.ws.rs.ext.ExceptionMapper;
+import javax.ws.rs.ext.Provider;
+
+@Provider
+public class NoPathExceptionMapper implements ExceptionMapper<NoPathParameterException> {
+    @Override
+    public Response toResponse(NoPathParameterException e) {
+        return Response.status(Response.Status.NOT_ACCEPTABLE).build();
```
Comment (tulinkry) [inline opengrok-web/src/main/java/org/opengrok/web/api/v1/NoPathExceptionMapper.java]: https://github.com/oracle/opengrok/pull/3050#discussion_r383695289

File: opengrok-indexer/src/main/java/org/opengrok/indexer/history/GitRepository.java (hunk 6 above)
Comment (vladak) [reply opengrok-indexer/src/main/java/org/opengrok/indexer/history/GitRepository.java]: Thanks guys. It occurred to me that this might be a problem but pushed the resolution aside. Will look into it.

File: opengrok-indexer/src/main/java/org/opengrok/indexer/history/Annotation.java [hunk 11]
```diff
@@ -163,10 +163,14 @@ void addLine(String revision, String author, boolean enabled) {
...
     void addDesc(String revision, String description) {
-        desc.put(revision, Util.encode(description));
+        desc.put(revision, description);
     }
 
     public String getDesc(String revision) {
+        return Util.encode(desc.get(revision));
+    }
+
+    public String getDescRaw(String revision) {
```
Comment (idodeclare) [inline opengrok-indexer/src/main/java/org/opengrok/indexer/history/Annotation.java]: Any possibility to make this class non-HTML aware and move the encoding to `-web` or `.web`?

Comment (vladak) [reply opengrok-indexer/src/main/java/org/opengrok/indexer/history/Annotation.java]: Sure; thought the `getDesc()` is used more widely.

File: opengrok-indexer/src/main/java/org/opengrok/indexer/web/Util.java [hunk 12]
```diff
@@ -713,82 +713,87 @@ public static void readableLine(int num, Writer out, Annotation annotation, Stri
...
+        if (enabled) {
+            out.write(anchorClassStart);
+            out.write("r");
+            out.write("\" style=\"background-color: ");
+            out.write(annotation.getColors().getOrDefault(r, "inherit"));
+            out.write("\" href=\"");
+            out.write(URIEncode(annotation.getFilename()));
+            out.write("?a=true&amp;r=");
+            out.write(URIEncode(r));
+            String msg = Util.encode(annotation.getDesc(r));
```
Comment (idodeclare) [inline opengrok-indexer/src/main/java/org/opengrok/indexer/web/Util.java]: Given the pre-existing `if (msg != null)` check nearby below, should perhaps the new `Util.encode()` be done in that if-block?

Comment (vladak) [reply opengrok-indexer/src/main/java/org/opengrok/indexer/web/Util.java]: definitely. Should have done the testing. Thanks.

File: opengrok-indexer/src/main/java/org/opengrok/indexer/history/GitRepository.java (hunk 6 above)
Comment (vladak) [reply opengrok-indexer/src/main/java/org/opengrok/indexer/history/GitRepository.java]: The latest changes have the basic `HeadHandler` (sans encoding).

File: opengrok-indexer/src/main/java/org/opengrok/indexer/util/HeadHandler.java [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+
+    private final List<String> lines = new ArrayList<>();
+
+    // TODO add encoding
+    public HeadHandler(int numlines) {
+        this.numlines = numlines;
+    }
+
+    public List<String> getLines() {
+        return lines;
```
Comment (idodeclare) [inline opengrok-indexer/src/main/java/org/opengrok/indexer/util/HeadHandler.java]: I recommend to avoid breaking encapsulation (though it's rampant in OpenGrok). E.g. rather than returning a private list, how about:

```
public int count() {
...
}

public String get(int index) {
...
}
```
```

File: opengrok-indexer/src/main/java/org/opengrok/indexer/util/HeadHandler.java [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+            if (line == null) { // EOF
+                return;
+            }
+            lines.add(line);
+            linenum++;
+        }
+        // Read and forget the rest. Need to use read() since it is not guaranteed
+        // that newline will ensue.
+        // TODO: would readline() get the job done ?
+        while (reader.read() != -1) {
```
Comment (idodeclare) [inline opengrok-indexer/src/main/java/org/opengrok/indexer/util/HeadHandler.java]: Technically this is doing needless `byte` to `char` conversion. E.g. were it a multi-byte encoding, then decoding is not worthwhile.

My original thinking on `HeadHandler` was to have the principal buffering at the byte-level. E.g. a `BufferedReader(InputStreamReader, /*tiny*/200)` atop a `BufferedInputStream`. Then when all the needed `readLine()` calls are done to just `read()` and discard from the `BufferedInputStream` — probably not a byte at a time but with say a very small `byte[1024]` buffer.

(Using `readLine()` here would defeat the goal of avoiding needless heap allocation.)

File: opengrok-indexer/src/main/java/org/opengrok/indexer/util/HeadHandler.java (hunk 13 above)
Comment (vladak) [reply opengrok-indexer/src/main/java/org/opengrok/indexer/util/HeadHandler.java]: I must admit OpenGrok code base has been primary source of my Java programming influence :-)

File: opengrok-indexer/src/main/java/org/opengrok/indexer/history/GitRepository.java [hunk 15]
```diff
@@ -389,23 +390,19 @@ private String getFirstRevision(String fullpath) throws IOException {
...
-                new InputStreamReader(executor.getOutputStream()))) {
-            String line;
+        HeadHandler headHandler = new HeadHandler(1);
+        int status = executor.exec(false, headHandler);
 
-            if ((line = in.readLine()) != null) {
-                return line.trim();
-            }
+        String line;
+        if ((line = headHandler.get(0)) != null) {
```
Comment (idodeclare) [inline opengrok-indexer/src/main/java/org/opengrok/indexer/history/GitRepository.java]: I think the API requires `if (headHandler.count() > 0) {`

Comment (vladak) [reply opengrok-indexer/src/main/java/org/opengrok/indexer/history/GitRepository.java]: I see, for some reason I thought it will return `null`.

File: opengrok-indexer/src/main/java/org/opengrok/indexer/util/HeadHandler.java (hunk 14 above)
Comment (vladak) [reply opengrok-indexer/src/main/java/org/opengrok/indexer/util/HeadHandler.java]: I see. Now, `BufferedReader.read()` will not read a single byte from the underlying stream - it will call `fill()` to fill the internal buffer.

File: opengrok-indexer/src/main/java/org/opengrok/indexer/history/GitRepository.java [hunk 16]
```diff
@@ -389,31 +390,24 @@ private String getFirstRevision(String fullpath) throws IOException {
...
-                new InputStreamReader(executor.getOutputStream()))) {
-            String line;
+        HeadHandler headHandler = new HeadHandler(1);
+        int status = executor.exec(false, headHandler);
 
-            if ((line = in.readLine()) != null) {
-                return line.trim();
-            }
+        String line;
+        if (headHandler.size() > 0 && (line = headHandler.get(0)) != null) {
```
Comment (idodeclare) [inline opengrok-indexer/src/main/java/org/opengrok/indexer/history/GitRepository.java]: Nit: I think your implementation does ensure non-null for a valid index. Maybe that could be stated firmly in the Javadoc?

File: opengrok-indexer/src/main/java/org/opengrok/indexer/util/HeadHandler.java (hunk 14 above)
Comment (vladak) [reply opengrok-indexer/src/main/java/org/opengrok/indexer/util/HeadHandler.java]: Anyhow, the suggested layering sounds good so I have implemented that, together with charset choice.

File: opengrok-indexer/src/main/java/org/opengrok/indexer/util/HeadHandler.java [hunk 17]
```diff
@@ -0,0 +1,98 @@
...
+     * @return line at given index
+     */
+    public String get(int i) {
+        return lines.get(i);
+    }
+
+    @Override
+    public void processStream(InputStream input) throws IOException {
+        BufferedInputStream bufStream = new BufferedInputStream(input);
+        BufferedReader reader = new BufferedReader(new InputStreamReader(bufStream, this.charset), 200);
```
[the remark this answers is not in the document]
Comment (vladak) [reply opengrok-indexer/src/main/java/org/opengrok/indexer/util/HeadHandler.java]: definitely. Wanted to do it but then forgot.

File: opengrok-indexer/src/main/java/org/opengrok/indexer/util/HeadHandler.java [hunk 18]
```diff
@@ -0,0 +1,98 @@
...
+            }
+            lines.add(line);
+            lineNum++;
+        }
+
+        // Read and forget the rest.
+        byte[] buf = new byte[1024];
+        int off = 0;
+        int len;
+        while ((len = bufStream.read(buf, off, buf.length)) != -1) {
```
[the remark this answers is not in the document]
Comment (vladak) [reply opengrok-indexer/src/main/java/org/opengrok/indexer/util/HeadHandler.java]: okay, sounds reasonable.

**Comment by vladak** [comment]
> LGTM

Thanks, this was nice learning experience.

File: opengrok-indexer/src/test/java/org/opengrok/indexer/util/HeadHandlerTest.java [hunk 19]
```diff
@@ -0,0 +1,91 @@
...
+
+        RandomInputStream rndStream = new RandomInputStream(totalCount, lines);
+        HeadHandler handler = new HeadHandler(headLines);
+        assertTrue(totalCount > handler.getBufferedReaderSize());
+        handler.processStream(rndStream);
+        assertEquals(headLines, handler.size());
+        assertEquals(totalCount + 1, rndStream.getCount());
+        int linesLen = 0;
+        for (int i = 0; i < headLines; i++) {
+            linesLen += handler.get(i).length();
```
Comment (idodeclare) [inline opengrok-indexer/src/test/java/org/opengrok/indexer/util/HeadHandlerTest.java]: With this approach using randomness, I wonder if it would be informative to compare the line content to what was generated.

Comment (tulinkry) [reply opengrok-indexer/src/test/java/org/opengrok/indexer/util/HeadHandlerTest.java]: For me it seems a little much work for a test; I'd prepare some files as resources (perhaps with this generator) and assert the content.

Comment (vladak) [reply opengrok-indexer/src/test/java/org/opengrok/indexer/util/HeadHandlerTest.java]: I wanted the test to adapt automatically based on changes to `HeadHandler` - in particular the internal buffer size.

File: opengrok-indexer/src/test/java/org/opengrok/indexer/util/HeadHandlerTest.java [hunk 20]
```diff
@@ -0,0 +1,91 @@
...
+            letters = letters + "0123456789";
+        }
+
+        public int getCount() {
+            return count;
+        }
+
+        @Override
+        public int read() throws IOException {
+            if (++count < maxcount) {
```
[the remark this answers is not in the document]
Comment (vladak) [reply opengrok-indexer/src/test/java/org/opengrok/indexer/util/HeadHandlerTest.java]: yes. was rough.

File: opengrok-indexer/src/test/java/org/opengrok/indexer/util/HeadHandlerTest.java [hunk 21]
```diff
@@ -0,0 +1,91 @@
...
+                // Want the newlines to appear within the first half.
+                if (count % ((maxcount / 2) / maxlines) == 0) {
```
[the remark this answers is not in the document]
Comment (vladak) [reply opengrok-indexer/src/test/java/org/opengrok/indexer/util/HeadHandlerTest.java]: good idea. I should randomize the test cases' parameters too.
